### DISCUSSION TITLE
CanAccessRecord() with CreateRule always returning false

### DISCUSSION THREAD

**Discussion by ahmeroxa** [pocketbase/pocketbase#3187] [Q&A] [answered]
I'm implementing RBAC with a fairly granular permissions system.

My create rule is as follows:
```
@request.auth.id != "" &&
@request.auth.roles.permission_set.entity_type ~ 'posts' &&
@request.auth.roles.permission_set.permissions.name ?= "create"
```

The requesting user has a relation to 1 or more `roles`, that have 1 or more relations to `permission_set` that includes an `entity_type` field (say `posts`) and a permission_set has a relation to 1 or more granular permissions e.g `list`, `create`, `read`, `update` and `delete`.

In the above case, the user must be associated with a role that includes a permission_set with the `entity_type` set to `posts` and at least a relation to `create`.

Unfortunately when trying to use `CanAccessRecord()` on create (and passing in the `CreateRule`), it always returns `false`.

I'm already using a similar pattern for all other operations and they work as expected.

I appreciate that with `create` the API employs a slightly more elaborate process with a `DrySubmit` but I would expect `CanAccessRecord()` to still work.

Here's a snippet of the code that I'm using to (try to) enforce the create rule:
```go
record := models.NewRecord(collection)
form := forms.NewRecordUpsert(app, record) // app is a core.App

form.LoadData(map[string]any{
	"title":   "foo",
	"author":   "me",
	"content": "content,
})

canAccess, err := r.app.Dao().CanAccessRecord(record, apis.RequestInfo(c), record.Collection().CreateRule)
if err != nil {
	return err
}
if !canAccess {
	return apis.NewForbiddenError("you do not have permission to create posts", nil)
} else {
	log.Println("user has access to create posts)
}
```

This is lifted pretty much straight out of the docs here:  https://pocketbase.io/docs/go-records/#check-if-record-can-be-accessed

Any idea what I'm doing wrong or how I can get it to work with `create`?

**Comment by ganigeorgiev** [comment, marked as answer]
In order the `CanAccessRecord` to work the record needs to exist in the database (internally the rule is converted to SQL statements).

This is also the reason why we perform `DrySubmit` with the regular record create action.

In your case you can try to wrap your code in a transaction and persist the record before performing the check or call `form.DrySubmit(txCallback)`, something like:

```go
...

err := form.DrySubmit(func(txDao *daos.Dao) error {
    canAccess, err := txDao.CanAccessRecord(record, apis.RequestInfo(c), record.Collection().CreateRule)
    if !canAccess {
        return apis.NewForbiddenError("you do not have permission to create posts", err)
    }
    return nil
})
if err != nil {
    return err
}


if err := form.Submit(); err != nil {
    return apis.NewBadRequestError("", err)
}

...
```

**Reply by ahmeroxa** [reply]
Got it! That works, thank you.
